drivers/meter/ade7978/ade7978.c

```diff
@@ -0,0 +1,576 @@
+	} else if ((reg_addr >= ADE7978_REG_PHSTATUS && reg_addr <= ADE7978_REG_CONFIG)
+		   ||
+		   (reg_addr >= ADE7978_REG_APF && reg_addr <= ADE7978_REG_LAST_RWDATA16) ||
+		   (reg_addr == ADE7978_REG_RUN)) {
+		/* 16 bits registers */
+		ret = no_os_spi_write_and_read(dev->spi_desc, buff, 5);
+		if (ret)
+			return ret;
+
+		*reg_data = no_os_get_unaligned_be16(&buff[3]);
+	} else {
+		/* 32 bits registers */
+		ret = no_os_spi_write_and_read(dev->spi_desc, buff, 7);
+		if (ret)
+			return ret;
+
+		*reg_data = no_os_get_unaligned_be32(&buff[3]);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Write device register.
+ * @param dev- The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The data to be written.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7978_write(struct ade7978_dev *dev, uint16_t reg_addr, uint32_t reg_data)
+{
+	/* data buffer */
+	uint8_t buff[7] = { 0 };
+	int i;
+
+	if (!dev)
+		return -ENODEV;
+	// write operation
+	buff[0] = ADE7978_SPI_WRITE_CMD;
+	no_os_put_unaligned_be16(reg_addr, &buff[1]);
+
+	/* 8 bits registers */
+	if ((reg_addr >= ADE7978_REG_MMODE && reg_addr <= ADE7978_REG_LAST_RWDATA8) ||
+	    (reg_addr >= ADE7978_REG_CONFIG2 && reg_addr <= ADE7978_SET_SPI_ADDR)) {
+		buff[3] = (uint8_t)reg_data;
+
+		return no_os_spi_write_and_read(dev->spi_desc, buff, 4);
+	}
+
+	/* 16 bits registers */
+	if ((reg_addr >= ADE7978_REG_PHSTATUS && reg_addr <= ADE7978_REG_CONFIG) ||
+	    (reg_addr >= ADE7978_REG_APF && reg_addr <= ADE7978_REG_LAST_RWDATA16) ||
+	    (reg_addr == ADE7978_REG_RUN)) {
+		no_os_put_unaligned_be16(reg_data, &buff[3]);
+
+		return no_os_spi_write_and_read(dev->spi_desc, buff, 5);
+	}
+
+	/* 32 bits registers */
+	no_os_put_unaligned_be32(reg_data, &buff[3]);
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff, 7);
+}
+
+/**
+ * @brief Update specific register bits.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Specific bits mask.
+ * @param reg_data - The data to be written.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7978_update_bits(struct ade7978_dev *dev, uint16_t reg_addr,
+			uint32_t mask, uint32_t reg_data)
+{
+	int ret;
+	/* register value read */
+	uint32_t data;
+
+	if (!dev)
+		return -ENODEV;
+
+	ret = ade7978_read(dev, reg_addr, &data);
+	if (ret)
+		return ret;
+
+	data &= ~mask;
+	data |= reg_data & mask;
+
+	return ade7978_write(dev, reg_addr, data);
+}
+
+/**
+ * @brief Get interrupt indicator from STATUS0 register.
+ * @param dev - The device structure.
+ * @param msk - Interrupt mask.
+ * @param status - Status indicator.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7978_get_int_status0(struct ade7978_dev *dev, uint32_t msk,
+			    uint8_t *status)
+{
+	int ret;
+	/* register value read */
+	uint32_t reg_val;
+
+	if (!dev)
+		return -ENODEV;
+	if (!status)
+		return -EINVAL;
```

Comment:
You can remove these, since they are checked in the `ade7978_read()` call below.